File: apriori/Apriori.php
<?php
class Apriori {
    /**
     * 最小支持度阈值
     * @var float
     */
    private $_support = 0.2;
    
    /**
     * 最小置信度阈值
     * @var float
     */
    private $_confidence = 0.5;
   
    /**
     * 数据项之间分隔符
     * @var string
     */
    private $_separation = '|';
    
    /**
     * 关联规则分隔符
     * @var string
     */
    private $_connection = '->';
    
    /**
     * 原始数据集
     * @var array
     */
    private $_data = [];
    
    /**
     * 保存唯一实例
     * @var Apriori
     */
    private static $_instance = null;
   
    /**
     * 禁止在类外通过调用构造函数实例化
     */
    private function construct() {
        
    }
    
    /**
     * 单例模式
     * @return Apriori
     */
    public static function getInstance() {
        if (is_null(self::$_instance)) {
            self::$_instance = new self();
        }
        return self::$_instance;
    }
    
    /**
     * 获得频繁项集
     * @return array
     */
    private function getFrequentCollections() {
        $frequents = [];
        $frequent = [];
        $count = count($this->_data);

        do {
            $candidate = $this->getCandidateCollections($frequent);

            $frequent = [];
            foreach ($candidate as $cand_k => $cand_v) {
                $cand_items = explode($this->_separation, $cand_k);
                foreach ($this->_data as $data) {
                    $items = explode($this->_separation, $data);
                    !array_diff($cand_items, $items) && $candidate[$cand_k] ++;
                }
                $candidate[$cand_k] /= $count;
            }
            $frequent = array_filter($candidate, function ($v) {
                return $v >= $this->_support;
            });

            if (empty($frequent)) {
                break;
            }
            $frequents[] = $frequent;            

        } while (true);

        return $frequents;
    }
    
    /**
     * 获得候选项集
     * @param array $frequent 频繁项集
     * @return array
     */
    private function getCandidateCollections(array $frequent) {
        $candidate = [];  
        $count = count($frequent);

        if ($count == 0) {
            foreach ($this->_data as $data) {
                $initial_data = explode($this->_separation, $data);
                foreach ($initial_data as $initial) {
                    !isset($candidate[$initial]) && $candidate[$initial] = 0;
                }
            }
            ksort($candidate);
            return $candidate;
        }
        
        $items = [];
        foreach ($frequent as $freq_k => $freq_v) {
            $items[] = explode($this->_separation, $freq_k);
        }

        for ($i = 0; $i < $count - 1; $i ++) {
            for ($j = $i + 1; $j < $count; $j ++) {
                $diff = array_diff($items[$i], $items[$j]);
                if (count($diff) != 1) {
                    continue;
                }
                $prepare = array_merge($items[$j], $diff);
                sort($prepare);
                $prepare_str = implode($this->_separation, $prepare);
                !isset($candidate[$prepare_str]) && $candidate[$prepare_str] = 0;
            }
        }  
        ksort($candidate);
        return $candidate;
    }
    
    /**
     * 获得真子集
     * @param array $set 集合项
     * @return array
     */
    private function getProperSubset(array $set) {
        $subset = [];
        $size = count($set);
        for ($i = 1; $i < $size; $i ++) {
            // $k_subset = [];
            $this->getKSubset($subset, $set, 0, $size, [], $i);
            // $subset[$i] = $k_subset;
        }
        return $subset;
    }
    
    /**
     * 递归获得k项真子集
     * @param array $k_subset
     * @param array $set
     * @param integer $start
     * @param integer $end
     * @param array $candidate
     * @param integer $k
     * @return void
     */
    private function getKSubset(array &$k_subset, array $set, $start, $end, array $candidate, $k) {
        if ($k == 0) {
            $k_subset[] = $candidate;
            return;
        }
        for ($i = $start; $i < $end; $i ++) {
            $copy = $candidate;
            $copy[] = $set[$i];
            $this->getKSubset($k_subset, $set, $i + 1, $end, $copy, $k - 1);
        }
    }
    
    /**
     * 非递归获得真子集
     * @param array $set
     * @return array
     */
    private function getSubset(array $set) {
        $pre_subset = [];
        $k_subset = [];

        foreach ($set as $data_k => $data_v) {
            $k_subset[] = [
                'next' => ++ $data_k,
                'value' => [$data_v]
            ];
        }

        $pre_subset = $k_subset;
        $count = count($set);
        $k = $count - 2;

        while ($k --) {
            $current = [];

            foreach ($k_subset as $item) {
                $c = $item['next'];

                while ($c < $count) {
                    $copy = $item['value'];
                    $copy[] = $set[$c];
                    $current[] = [
                        'next' => $c ++,
                        'value' => $copy
                    ];
                }
            }
            $k_subset = $current;
            $pre_subset = array_merge($pre_subset, $k_subset);
        }
        return array_map(function ($v) {
            return $v['value'];
        }, $pre_subset);
    }
    /*
    // 非递归获得真子集，二进制方式
    private function getProperSubsetUnrecursive($set) {
        $total = 1;
        $size = count($set);
        $subset = [];
        for ($n = 0; $n ++ < $size; $total <<= 1);
        for ($i = 1; $i < $total; $i ++) {
            $mask = $i;
            $subset[] = array_filter($set, function ($v) use (&$mask) {
                $flag = $mask & 1 == 1;
                $mask >>= 1;
                return $flag;
            });
        }
        return $subset;
    }*/
   
    /**
     * 获得关联规则
     * @param array $primary_data
     * @param float $support
     * @param float $confidence
     * @return number[][]|mixed[][]
     */
    public function getRelationRules(array $primary_data, $support, $confidence) {
        $data = [];
        foreach ($primary_data as $v) {
            $data[] = implode($this->_separation, $v);
        }
        $this->_data = $data;
        $this->_support = floatval($support);
        $this->_confidence = floatval($confidence);
        
        $relation = [];
        $frequents = $this->getFrequentCollections();
        $count = count($frequents);

        for ($i = 1; $i < $count; $i ++) {
            $frequent = $frequents[$i];

            foreach ($frequent as $freq_k => $freq_v) {
                $freq = explode($this->_separation, $freq_k);
                $subset = $this->getProperSubset($freq);

                foreach ($subset as $sub_item) {
                    $sub_item_str = implode($this->_separation, $sub_item);
                    $target_str = implode($this->_separation, array_diff($freq, $sub_item));
                    $prepare_str = implode($this->_connection, [$sub_item_str, $target_str]);

                    if (!isset($relation[$prepare_str])) {
                        $confidence_v = round($freq_v / $frequents[count($sub_item) - 1][$sub_item_str], 4);
                        $confidence_v >= $this->_confidence && $relation[$prepare_str] = [
                            'support' => $freq_v,
                            'confidence' => $confidence_v
                        ];
                    }
                }
            }
        }

        return $relation;
    }
}

File: apriori/apriori.py
#!/usr/bin/python
#coding: UTF-8
import os


class Apriori:
    

    def __init__(self, data, support, confidence):
        self.__separation = '|'
        self.__connection = '->'
        self.__support = float(support) 
        self.__confidence = float(confidence)
        self.__data = data

    
    def __getFrequentCollections(self):
        frequents = []
        frequent = {}
        count = len(self.__data)

        while True:
            candidate = self.__getCandidateCollections(frequent)
            frequent = {}
            for candK, candV in candidate.items():
                candItems = candK.split(self.__separation)

                for items in self.__data:
                    # list(set(candItems).difference(set(items)))
                    if not [v for v in candItems if v not in items]:
                        candidate[candK] += 1

                candidate[candK] /= count * 1.0

            frequent = dict(filter(lambda (k, v): v >= self.__support, candidate.items()))

            if not frequent:
                break

            frequents.append(frequent)

        return frequents


    def __getCandidateCollections(self, frequent):
        tmp = []
        count = len(frequent)

        if count == 0:
            for k in self.__data:
                tmp.extend(k)
            tmp = list(set(tmp))
                        
        else:
            items = [k.split(self.__separation) for k, v in frequent.items()]

            for i in range(count - 1):
                for j in range(count):
                    diff = [v for v in items[i] if v not in items[j]]
                    if len(diff) != 1:
                        continue

                    prepare = sorted(items[j] + diff)
                    prepareStr = self.__separation.join(prepare)
                    if prepareStr not in tmp:
                        tmp.append(prepareStr)

        tmp.sort()
        candidate = {k: 0 for k in tmp}
        return candidate

   
    def __getProperSubset(self, sets):
        return [[v for k, v in enumerate(sets) if (i >> k) & 1 == 1] for i in range(1, (1 << len(sets)) - 1)]


    def getRelationRules(self):
        relation = {}
        frequents = self.__getFrequentCollections();

        for frequent in frequents:

            for freqK, freqV in frequent.items():
                freq = freqK.split(self.__separation)
                subset = self.__getProperSubset(freq)
    
                for subItem in subset:
                    subItemStr = self.__separation.join(subItem)
                    targetStr = self.__separation.join([v for v in freq if v not in subItem])
                    prepareStr = self.__connection.join([subItemStr, targetStr])
    
                    if prepareStr not in relation.keys():
                        confidenceV = freqV / frequents[len(subItem) - 1][subItemStr]
                        if confidenceV >= self.__confidence:
                            relation[prepareStr] = {
                                'support': freqV,
                                'confidence': confidenceV
                            }
        
        return relation



if __name__ == '__main__':
    data = []
    with open('data.txt') as f:
        data = [l.strip().split(',') for l in f if l]

    apr = Apriori(data, 0.2, 0.5)
    relation = apr.getRelationRules()

    print(relation)
